File: backend/src/schemas/budget.py
"""Budget and cost tracking schemas."""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class CostRecordCreate(BaseModel):
    agent_id: uuid.UUID
    task_id: Optional[uuid.UUID] = None
    amount_cents: int
    token_count: Optional[int] = None
    model_name: Optional[str] = None


class CostRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: uuid.UUID
    agent_id: uuid.UUID
    task_id: Optional[uuid.UUID] = None
    amount_cents: int
    token_count: Optional[int] = None
    model_name: Optional[str] = None
    recorded_at: datetime


class AgentBudgetSummary(BaseModel):
    agent_id: uuid.UUID
    agent_name: str
    monthly_budget_cents: Optional[int] = None
    current_month_spent_cents: int = 0
    budget_remaining_cents: Optional[int] = None
    utilization_pct: Optional[float] = None


class BudgetOverviewResponse(BaseModel):
    total_budget_cents: int
    total_spent_cents: int
    total_remaining_cents: int
    agent_summaries: list[AgentBudgetSummary]

File: frontend/src/types/budget.ts
export interface AgentBudgetSummary {
  agent_id: string
  agent_name: string
  monthly_budget_cents: number | null
  current_month_spent_cents: number
  budget_remaining_cents: number | null
  utilization_pct: number | null
}

export interface CostRecord {
  id: string
  tenant_id: string
  agent_id: string
  task_id: string | null
  amount_cents: number
  token_count: number | null
  model_name: string | null
  recorded_at: string
}

export interface BudgetOverview {
  total_budget_cents: number
  total_spent_cents: number
  total_remaining_cents: number
  agent_summaries: AgentBudgetSummary[]
}
